### PR TITLE
address some of #33: more COSE format tests

### PR DESCRIPTION
**Pull request by mozkeeler**

### CODE REVIEW

**Comment by mozkeeler** [comment]
I tried to go through and test every aspect of how some data could be valid CBOR but not valid COSE for our purposes. Let me know what you think.

File: src/decoder.rs [hunk 1]
```diff
@@ -103,9 +114,12 @@ fn decode_signature_struct(
         _ => return Err(CoseError::UnexpectedType),
     };
 
-    let ee_cert = &get_map_value(&protected_signature_header, &CborType::Integer(4))?;
+    let ee_cert = &get_map_value(protected_signature_header, &CborType::Integer(4))?;
     let ee_cert = unpack!(Bytes, ee_cert).clone();
 
+    // The unprotected header section is expected to be an empty map.
+    validate_empty_map(&cose_signature[1])?;
```
Comment (franziskuskiefer) [inline src/decoder.rs]: Maybe `is_empty_map`? Not sure what I expect `validate_` to do but not necessarily to check whether it's an empty map or not.

File: src/test_cose.rs [hunk 2]
```diff
@@ -236,3 +238,481 @@ fn test_cose_sign_verify_two_signatures_tampered_signature() {
...
+        vec![CborType::Array(Vec::new())],
+    )
+}
+
+// Helper function to create a minimally-valid COSE_Signature (i.e. "body").
+fn make_minimally_valid_cose_signature_protected_header() -> Vec<u8> {
+    encode_test_protected_header(
+        vec![CborType::Integer(1),
+             CborType::Integer(4)],
+        vec![CborType::SignedInteger(-7),
```
Comment (franziskuskiefer) [inline src/test_cose.rs]: `COSE_TYPE_ES256` for `-7`.

File: src/test_cose.rs [hunk 3]
```diff
@@ -236,3 +238,481 @@ fn test_cose_sign_verify_two_signatures_tampered_signature() {
...
+    let payload = vec![0];
+    let result = decode_signature(bytes, &payload);
+    assert!(result.is_err());
+    assert_eq!(result.err(), Some(expected_error));
+}
+
+// Helper function to take a `Vec<CborType>`, wrap it in a `CborType::Array`, tag it with the
+// COSE_Sign tag (98), and serialize it to a `Vec<u8>`.
+fn wrap_tag_and_encode_array(array: Vec<CborType>) -> Vec<u8> {
+    CborType::Tag(98, Box::new(CborType::Array(array))).serialize()
```
Comment (franziskuskiefer) [inline src/test_cose.rs]: `COSE_SIGN_TAG` for `98`.

File: src/test_cose.rs [hunk 4]
```diff
@@ -236,3 +238,481 @@ fn test_cose_sign_verify_two_signatures_tampered_signature() {
...
+fn build_test_cose_signature(protected_header: Vec<u8>) -> CborType {
+    CborType::Array(vec![CborType::Bytes(protected_header),
+         CborType::Map(BTreeMap::new()),
+         CborType::Bytes(Vec::new())])
+}
+
+// Helper function to create the minimally-valid COSE_Sign (i.e. "body") protected header.
+fn make_minimally_valid_cose_sign_protected_header() -> Vec<u8> {
+    encode_test_protected_header(
+        vec![CborType::Integer(4)],
```
Comment (franziskuskiefer) [inline src/test_cose.rs]: We don't have a `COSE_KID` value but we should maybe introduce one.

File: src/test_cose.rs [hunk 5]
```diff
@@ -3,6 +3,8 @@ use util_test::{sign, verify_signature};
 use {CoseError, SignatureAlgorithm, SignatureParameters};
```
Comment (franziskuskiefer) [inline src/test_cose.rs]: This file is getting a little large. Maybe we should separate between COSE sign/verify tests and decoder tests. When we fix #46 the sign/verify tests should go away.

File: src/test_cose.rs [hunk 6]
```diff
@@ -236,3 +238,481 @@ fn test_cose_sign_verify_two_signatures_tampered_signature() {
...
+    let values = vec![CborType::Bytes(body_protected_header),
+                      CborType::Map(BTreeMap::new()),
+                      CborType::Null,
+                      CborType::Array(Vec::new())];
+    let bytes = wrap_tag_and_encode_array(values);
+    test_cose_format_error(&bytes, CoseError::MalformedInput);
+}
+
+#[test]
+fn test_cose_signature_protected_header_wrong_type() {
```
Comment (franziskuskiefer) [inline src/test_cose.rs]: I'm sure there are some more parts you could pull out to not duplicate the code. But that might make it only harder to read.

File: src/decoder.rs (hunk 1 above)
Comment (mozkeeler) [reply src/decoder.rs]: Ok - I changed the name and added some documentation.

File: src/test_cose.rs (hunk 5 above)
Comment (mozkeeler) [reply src/test_cose.rs]: It's just a test file, though - hopefully we won't even have to open it in the future.

File: src/test_cose.rs (hunk 3 above)
Comment (mozkeeler) [reply src/test_cose.rs]: Sounds good.

File: src/test_cose.rs (hunk 4 above)
Comment (mozkeeler) [reply src/test_cose.rs]: Sounds good.

File: src/test_cose.rs (hunk 2 above)
Comment (mozkeeler) [reply src/test_cose.rs]: Replaced.

File: src/test_cose.rs (hunk 6 above)
Comment (mozkeeler) [reply src/test_cose.rs]: Yeah, I don't think it would be that beneficial.

**Pull request merged**